### PR TITLE
[Merged by Bors] - Wasm optimization tips

### PR DESCRIPTION
**Pull request by nicopap**
# Objective

Add a section to the example's README on how
to reduce generated wasm executable size.

Add a `wasm-release` profile to bevy's `Cargo.toml`
in order to use it when building bevy-website.

Notes:
- We do not recommend `strip = "symbols"` since it breaks bindgen
- see https://github.com/bevyengine/bevy-website/pull/402

### CODE REVIEW

File: examples/README.md [hunk 1]
```diff
@@ -480,6 +482,72 @@ python3 -m http.server --directory examples/wasm
...
 
+### Optimizing
+
+On the web, it's useful to reduce the size of the files that are distributed.
+With rust, there are many ways to improve your executable sizes.
+Here are some.
+
+#### 1. Tweak your `Cargo.toml`
+
+Add this to your `Cargo.toml`:
```
Comment (mockersf) [inline examples/README.md:493]: this part could use a link to https://doc.rust-lang.org/cargo/reference/profiles.html

File: examples/README.md [hunk 2]
```diff
@@ -480,6 +482,72 @@ python3 -m http.server --directory examples/wasm
...
+
+```toml
+[profile.release]
```
Comment (mockersf) [inline examples/README.md:496]: I would rather have this use another profile name than `release`

**Comment by mockersf** [comment]
Could you also add a new profile to the root `Cargo.toml`?

```toml
[profile.wasm-release]
inherits = "release"
opt-level = "z"
lto = "fat"
codegen-units = 1
```

and showing how to use it (`--profile wasm-release` instead of `--release`)

**Review by mockersf — APPROVED** [review]

**Comment by mockersf** [comment]
adding the 0.8 milestone as otherwise we would have to wait for the next release to use this on the website

**Comment by nicopap** [comment]
I guess the new profile can always be split from the doc update, if push comes to shove.

**Comment by nicopap** [comment]
There is also space for improvement in the size optimization tips. For example, removing `DefaultPlugins` and only adding the ones necessary, etc. But the ones I gave are basically "free" and relevant to all rust projects.

Maybe it would be worth creating more bevy specific optim instructions in a potential bevy guide/manual and link to that page in the future.

**Comment by mockersf** [comment]
> There is also space for improvement in the size optimization tips. For example, removing `DefaultPlugins` and only adding the ones necessary, etc. But the ones I gave are basically "free" and relevant to all rust projects.

No, keep using `DefaultPlugins`, but disable default features ob bevy and only enable those you actually uses 😉 

> Maybe it would be worth creating more bevy specific optim instructions in a potential bevy guide/manual and link to that page in the future.

Definitely! We need a good guide on "how to build something optimised and ready for release". https://github.com/bevyengine/bevy/issues/1885

**Comment by colepoirier** [comment]
How much slower were the binraries produced with opt-level=z compared to normal —release optimization?

**Comment by nicopap** [comment]
You can run the examples with the changeset from https://github.com/bevyengine/bevy-website/pull/402

Remember, those optimizations are only for the wasm examples available on the bevy website, it is **much better** user experience to have code that runs half as fast and loads two times faster. Especially given none of the examples are especially taxing.

Regardless, I tested with the `release` profile, and switching `wasm-opt` to `-O3` and I ran bevymark, went from
- `-Oz`: `85K` at 20fps
- `-O3`: `133K` at 20fps

To me, it means we should go with this change.

**Comment by colepoirier** [comment]
> Regardless, I tested with the `release` profile, and switching `wasm-opt` to `-O3` and I ran bevymark, went from
> 
> * `-Oz`: `85K` at 20fps
> * `-O3`: `133K` at 20fps
> 
> To me, it means we should go with this change.

Thanks for running the numbers, looks good to me, I agree with you. I also agree with the other things you said in the comment that I haven‘t quoted in this reply.

**Comment by rparrett** [comment]
Would it be possible / desirable to use the normal profile specifically for the stress tests?

**Comment by nicopap** [comment]
> Would it be possible / desirable to use the normal profile specifically for the stress tests?

Actually quite a good idea. Fairly easy to implement too.

**Pull request closed without merging**